File: src/mediaKeywordSelection.ts
export type MediaKeywordCandidate = {
  keyword: string;
  volume?: number;
  intent?: string;
  cpc?: number;
  competition?: string;
  category?: string;
  relevanceScore?: number;
  suggestedPhrase?: string;
};

export type MediaKeywordSelectionInput = {
  coreKeyword: string;
  selectedCategory?: string;
  sourceText?: string;
  keywords: MediaKeywordCandidate[];
  limit?: number;
};

export type MediaSeoKeywordOutput = {
  filename: string;
  title: string;
  alt_text: string;
  caption: string;
  description: string;
};

export type MediaKeywordUsage = {
  coreKeyword: string;
  candidateKeywords: string[];
  usedKeywords: string[];
  warnings: string[];
  validationStatus: 'passed' | 'core-only' | 'inferred';
};

const MEDIA_KEYWORD_STOP_WORDS = new Set([
  'and', 'for', 'from', 'image', 'images', 'photo', 'photos', 'product', 'products',
  'the', 'this', 'with',
]);

export const normalizeMediaKeywordPhrase = (value: unknown): string => (
  String(value ?? '')
    .normalize('NFKC')
    .toLowerCase()
    .replace(/[^a-z0-9\u4e00-\u9fff]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
);

const keywordTokens = (value: unknown) => (
  normalizeMediaKeywordPhrase(value)
    .split(' ')
    .filter(token => token.length > 1 && !MEDIA_KEYWORD_STOP_WORDS.has(token))
);

const sharedTokenCount = (left: string[], right: string[]) => {
  const rightSet = new Set(right);
  return left.reduce((count, token) => count + (rightSet.has(token) ? 1 : 0), 0);
};

export const selectMediaKeywordCandidates = ({
  coreKeyword,
  selectedCategory = '',
  sourceText = '',
  keywords,
  limit = 12,
}: MediaKeywordSelectionInput): MediaKeywordCandidate[] => {
  const normalizedCore = normalizeMediaKeywordPhrase(coreKeyword);
  const normalizedCategory = normalizeMediaKeywordPhrase(selectedCategory).replace(/\s+/g, '-');
  const coreTokens = keywordTokens(coreKeyword);
  const sourceTokens = keywordTokens(sourceText);
  const seen = new Set<string>();
  const ranked: Array<{ item: MediaKeywordCandidate; score: number; index: number }> = [];

  keywords.forEach((item, index) => {
    const keyword = String(item?.keyword ?? '').replace(/\s+/g, ' ').trim();
    const normalizedKeyword = normalizeMediaKeywordPhrase(keyword);
    if (!normalizedKeyword || normalizedKeyword === normalizedCore || seen.has(normalizedKeyword)) return;
    seen.add(normalizedKeyword);

    const normalizedItemCategory = normalizeMediaKeywordPhrase(item.category).replace(/\s+/g, '-');
    const categoryMatch = Boolean(normalizedCategory && normalizedItemCategory === normalizedCategory);
    const containsCore = Boolean(normalizedCore && normalizedKeyword.includes(normalizedCore));
    const itemTokens = keywordTokens(keyword);
    const coreOverlap = sharedTokenCount(itemTokens, coreTokens);
    const sourceOverlap = sharedTokenCount(itemTokens, sourceTokens);
    if (!categoryMatch && !containsCore && coreOverlap === 0 && sourceOverlap === 0) return;

    const relevance = Number.isFinite(Number(item.relevanceScore)) ? Number(item.relevanceScore) : 0;
    const volume = Number.isFinite(Number(item.volume)) ? Math.max(0, Number(item.volume)) : 0;
    const score = (
      (categoryMatch ? 1000 : 0)
      + (containsCore ? 500 : 0)
      + coreOverlap * 50
      + sourceOverlap * 20
      + relevance
      + Math.min(20, Math.log10(volume + 1) * 5)
    );
    ranked.push({ item: { ...item, keyword }, score, index });
  });

  const cappedLimit = Math.max(1, Math.min(50, Math.trunc(limit || 12)));
  return ranked
    .sort((left, right) => right.score - left.score || left.index - right.index)
    .slice(0, cappedLimit)
    .map(entry => entry.item);
};

const includesNormalizedPhrase = (text: unknown, phrase: unknown) => {
  const normalizedText = normalizeMediaKeywordPhrase(text);
  const normalizedPhrase = normalizeMediaKeywordPhrase(phrase);
  return Boolean(normalizedPhrase && normalizedText.includes(normalizedPhrase));
};

export const validateMediaKeywordUsage = ({
  coreKeyword,
  candidates,
  output,
}: {
  coreKeyword: string;
  candidates: MediaKeywordCandidate[];
  output: MediaSeoKeywordOutput;
}): MediaKeywordUsage => {
  const cleanCoreKeyword = String(coreKeyword || '').replace(/\s+/g, ' ').trim();
  const fieldLimits: Array<[keyof MediaSeoKeywordOutput, number]> = [
    ['filename', 125],
    ['title', 60],
    ['alt_text', 125],
    ['caption', 120],
    ['description', 160],
  ];
  for (const [field, limit] of fieldLimits) {
    const value = String(output[field] || '').trim();
    if (!value || value.length > limit) {
      throw new Error(`Generated ${field} failed the required length limit (${limit})`);
    }
  }
  const supportingFields = [output.alt_text, output.caption, output.description]
    .map(normalizeMediaKeywordPhrase);
  if (new Set(supportingFields).size === 1) {
    throw new Error('Generated SEO supporting fields are overly repetitive');
  }
  if (cleanCoreKeyword && (
    !includesNormalizedPhrase(output.filename, cleanCoreKeyword)
    || !includesNormalizedPhrase(output.title, cleanCoreKeyword)
  )) {
    throw new Error('Generated filename and title must include the core keyword');
  }

  const candidateKeywords = Array.from(new Map(
    candidates
      .map(item => String(item.keyword || '').replace(/\s+/g, ' ').trim())
      .filter(Boolean)
      .map(keyword => [normalizeMediaKeywordPhrase(keyword), keyword] as const),
  ).values());
  const supportingText = [output.alt_text, output.caption, output.description].join(' ');
  const usedKeywords = candidateKeywords.filter(keyword => includesNormalizedPhrase(supportingText, keyword));

  if (candidateKeywords.length > 0 && usedKeywords.length === 0) {
    throw new Error('Generated SEO did not use a supporting keyword from the uploaded table');
  }
  if (usedKeywords.length > 3) {
    throw new Error('Generated SEO used more than three supporting keywords');
  }

  return {
    coreKeyword: cleanCoreKeyword,
    candidateKeywords,
    usedKeywords,
    warnings: candidateKeywords.length > 0 ? [] : (cleanCoreKeyword ? ['词表无匹配词'] : []),
    validationStatus: candidateKeywords.length > 0 ? 'passed' : (cleanCoreKeyword ? 'core-only' : 'inferred'),
  };
};

export const generateValidatedMediaSeo = async ({
  coreKeyword,
  candidates,
  generate,
  maxAttempts = 2,
}: {
  coreKeyword: string;
  candidates: MediaKeywordCandidate[];
  generate: (context: {
    attempt: number;
    validationFeedback: string;
  }) => Promise<MediaSeoKeywordOutput>;
  maxAttempts?: number;
}): Promise<{ output: MediaSeoKeywordOutput; keywordUsage: MediaKeywordUsage }> => {
  const attempts = Math.max(1, Math.min(2, Math.trunc(maxAttempts || 2)));
  let validationFeedback = '';

  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    const output = await generate({ attempt, validationFeedback });
    try {
      const keywordUsage = validateMediaKeywordUsage({ coreKeyword, candidates, output });
      return { output, keywordUsage };
    } catch (error) {
      validationFeedback = error instanceof Error ? error.message : String(error);
      if (attempt === attempts) {
        throw new Error(`Media SEO validation failed after ${attempts} attempts: ${validationFeedback}`);
      }
    }
  }

  throw new Error(`Media SEO validation failed after ${attempts} attempts`);
};

const normalizeMediaSeoFilename = (value: unknown) => {
  const withoutExtension = String(value || '').replace(/\.[a-z0-9]+$/i, '');
  const stem = normalizeMediaKeywordPhrase(withoutExtension).replace(/\s+/g, '-') || 'image';
  return `${stem}.webp`;
};

export const reserveUniqueMediaSeoFilename = (
  proposedFilename: string,
  mediaId: number,
  usedFilenames: Set<string>,
  sourceHint = '',
) => {
  const normalized = normalizeMediaSeoFilename(proposedFilename);
  const normalizedKey = normalized.toLowerCase();
  if (!usedFilenames.has(normalizedKey)) {
    usedFilenames.add(normalizedKey);
    return normalized;
  }

  const stem = normalized.replace(/\.webp$/i, '');
  const stemTokens = new Set(stem.split('-'));
  const hintTokens = normalizeMediaKeywordPhrase(sourceHint).split(' ').filter(token => (
    token.length > 1 && !stemTokens.has(token) && !MEDIA_KEYWORD_STOP_WORDS.has(token)
  ));
  const detailTerms = new Set(['front', 'rear', 'back', 'side', 'top', 'detail', 'closeup', 'angle', 'scene', 'installed', 'indoor', 'outdoor']);
  const prioritizedHints = hintTokens.filter(token => /\d/.test(token) || detailTerms.has(token));
  const selectedHints = [...prioritizedHints, ...hintTokens]
    .filter((token, index, all) => all.indexOf(token) === index)
    .slice(0, 2);
  if (selectedHints.length) {
    const detailed = `${stem}-${selectedHints.join('-')}.webp`;
    if (!usedFilenames.has(detailed.toLowerCase())) {
      usedFilenames.add(detailed.toLowerCase());
      return detailed;
    }
  }
  let suffix = String(Math.max(1, Math.trunc(Number(mediaId) || 0)));
  let unique = `${stem}-${suffix}.webp`;
  let counter = 2;
  while (usedFilenames.has(unique.toLowerCase())) {
    suffix = `${Math.max(1, Math.trunc(Number(mediaId) || 0))}-${counter}`;
    unique = `${stem}-${suffix}.webp`;
    counter += 1;
  }
  usedFilenames.add(unique.toLowerCase());
  return unique;
};

File: backend/media_keyword_selection.py
import math
import re
import unicodedata
from typing import Any


_STOP_WORDS = {"and", "for", "from", "image", "images", "photo", "photos", "product", "products", "the", "this", "with"}


def normalize_phrase(value: Any) -> str:
    text = unicodedata.normalize("NFKC", str(value or "")).lower()
    return re.sub(r"\s+", " ", re.sub(r"[^a-z0-9\u4e00-\u9fff]+", " ", text)).strip()


def _tokens(value: Any) -> list[str]:
    return [token for token in normalize_phrase(value).split(" ") if len(token) > 1 and token not in _STOP_WORDS]


def select_keyword_candidates(
    core_keyword: str,
    selected_category: str,
    source_text: str,
    keywords: list[dict[str, Any]],
    limit: int = 12,
) -> list[dict[str, Any]]:
    core = normalize_phrase(core_keyword)
    category = normalize_phrase(selected_category).replace(" ", "-")
    core_tokens = set(_tokens(core_keyword))
    source_tokens = set(_tokens(source_text))
    seen: set[str] = set()
    ranked: list[tuple[float, int, dict[str, Any]]] = []
    for index, row in enumerate(keywords or []):
        if not isinstance(row, dict):
            continue
        keyword = re.sub(r"\s+", " ", str(row.get("keyword") or "")).strip()
        normalized = normalize_phrase(keyword)
        if not normalized or normalized == core or normalized in seen:
            continue
        seen.add(normalized)
        row_category = normalize_phrase(row.get("category")).replace(" ", "-")
        category_match = bool(category and category == row_category)
        contains_core = bool(core and core in normalized)
        row_tokens = set(_tokens(keyword))
        core_overlap = len(row_tokens & core_tokens)
        source_overlap = len(row_tokens & source_tokens)
        if not category_match and not contains_core and not core_overlap and not source_overlap:
            continue
        try:
            relevance = float(row.get("relevanceScore") or 0)
        except (TypeError, ValueError):
            relevance = 0
        try:
            volume = max(0.0, float(row.get("volume") or 0))
        except (TypeError, ValueError):
            volume = 0
        score = (1000 if category_match else 0) + (500 if contains_core else 0)
        score += core_overlap * 50 + source_overlap * 20 + relevance + min(20, math.log10(volume + 1) * 5)
        ranked.append((score, index, {**row, "keyword": keyword}))
    ranked.sort(key=lambda entry: (-entry[0], entry[1]))
    return [entry[2] for entry in ranked[:max(1, min(12, int(limit or 12)))]]


def _includes(text: Any, phrase: Any) -> bool:
    normalized_phrase = normalize_phrase(phrase)
    return bool(normalized_phrase and normalized_phrase in normalize_phrase(text))


def validate_keyword_usage(
    core_keyword: str,
    candidates: list[dict[str, Any]],
    output: dict[str, Any],
) -> dict[str, Any]:
    core = re.sub(r"\s+", " ", str(core_keyword or "")).strip()
    for field, limit in (("filename", 125), ("title", 60), ("alt_text", 125), ("caption", 120), ("description", 160)):
        value = str(output.get(field) or "").strip()
        if not value or len(value) > limit:
            raise ValueError(f"Generated {field} failed the required length limit ({limit})")
    supporting_fields = [normalize_phrase(output.get(field)) for field in ("alt_text", "caption", "description")]
    if len(set(supporting_fields)) == 1:
        raise ValueError("Generated SEO supporting fields are overly repetitive")
    if core and (not _includes(output.get("filename"), core) or not _includes(output.get("title"), core)):
        raise ValueError("Generated filename and title must include the core keyword")
    candidate_keywords: list[str] = []
    seen: set[str] = set()
    for row in candidates or []:
        keyword = re.sub(r"\s+", " ", str(row.get("keyword") or "")).strip()
        normalized = normalize_phrase(keyword)
        if keyword and normalized not in seen:
            seen.add(normalized)
            candidate_keywords.append(keyword)
    supporting_text = " ".join(str(output.get(field) or "") for field in ("alt_text", "caption", "description"))
    used = [keyword for keyword in candidate_keywords if _includes(supporting_text, keyword)]
    if candidate_keywords and not used:
        raise ValueError("Generated SEO did not use a supporting keyword from the uploaded table")
    if len(used) > 3:
        raise ValueError("Generated SEO used more than three supporting keywords")
    return {
        "coreKeyword": core,
        "candidateKeywords": candidate_keywords,
        "usedKeywords": used,
        "warnings": [] if candidate_keywords else (["词表无匹配词"] if core else []),
        "validationStatus": "passed" if candidate_keywords else ("core-only" if core else "inferred"),
    }
